xdsl/dialects/arith.py

```diff
@@ -148,21 +149,55 @@ class ConstantOp(IRDLOperation):
     @overload
     def __init__(
         self,
-        value: IntegerAttr | FloatAttr[AnyFloat] | DenseIntOrFPElementsAttr,
-        value_type: None = None,
+        value: IntegerAttr,
+        value_type: IntegerType | None = None,
     ) -> None: ...
 
     @overload
-    def __init__(self, value: Attribute, value_type: Attribute) -> None: ...
+    def __init__(
+        self,
+        value: FloatAttr[AnyFloat],
+        value_type: AnyFloat | None = None,
+    ) -> None: ...
 
+    @overload
     def __init__(
         self,
-        value: IntegerAttr | FloatAttr[AnyFloat] | Attribute,
+        value: DenseIntOrFPElementsAttr,
+        value_type: RankedStructure[AnyFloat | IntegerType | IndexType] | None = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self,
+        value: Attribute,
+        value_type: Attribute,
+    ) -> None: ...
+
+    def __init__(
+        self,
+        value: IntegerAttr | FloatAttr[AnyFloat] | DenseIntOrFPElementsAttr | Attribute,
         value_type: Attribute | None = None,
     ):
         if value_type is None:
-            value = cast(IntegerAttr | FloatAttr[AnyFloat], value)
-            value_type = value.type
+            if isinstance(value, IntegerAttr | FloatAttr):
+                value = cast(IntegerAttr | FloatAttr[AnyFloat], value)
+                value_type = value.type
+            elif isinstance(value, DenseIntOrFPElementsAttr):
+                value_type = value.type
```

Comment:
You should use `value.get_type()` to get the type associated to a `TypedAttribute`

Reply:
done